core/src/main/java/org/apache/iceberg/SnapshotManager.java

```diff
@@ -19,292 +19,48 @@
-        isCurrentAncestor(current, snapshotId),
-        "Cannot roll back to snapshot, not an ancestor of the current state: %s", snapshotId);
-    return setCurrentSnapshot(snapshotId);
-  }
-
-  @Override
-  public Object updateEvent() {
-    if (targetSnapshotId == null) {
-      // NOOP operation, no snapshot created
-      return null;
-    }
-
-    switch (managerOperation) {
-      case ROLLBACK:
-        // rollback does not create a new snapshot
-        return null;
-      case CHERRYPICK:
-        TableMetadata tableMetadata = refresh();
-        long snapshotId = tableMetadata.currentSnapshot().snapshotId();
-        if (targetSnapshotId == snapshotId) {
-          // No new snapshot is created for fast-forward
-          return null;
-        } else {
-          // New snapshot created, we rely on super class to fire a CreateSnapshotEvent
-          return super.updateEvent();
-        }
-      default:
-        throw new UnsupportedOperationException(managerOperation + " is not supported");
-    }
-  }
-
-  @Override
-  protected void validate(TableMetadata base) {
-    validateCurrentSnapshot(base, requiredCurrentSnapshotId);
-    validateNonAncestor(base, targetSnapshotId);
-    validateReplacedPartitions(base, overwriteParentId, replacedPartitions);
-    WapUtil.validateWapPublish(base, targetSnapshotId);
+    transaction.setBranchSnapshot().rollbackTo(snapshotId).commit();
+    return this;
   }
 
   @Override
   public Snapshot apply() {
-    TableMetadata base = refresh();
-
-    if (targetSnapshotId == null) {
-      // if no target snapshot was configured then NOOP by returning current state
-      return base.currentSnapshot();
-    }
-
-    switch (managerOperation) {
-      case CHERRYPICK:
-        if (base.snapshot(targetSnapshotId).parentId() != null &&
-            base.currentSnapshot().snapshotId() == base.snapshot(targetSnapshotId).parentId()) {
-          // the snapshot to cherrypick is already based on the current state: fast-forward
-          validate(base);
-          return base.snapshot(targetSnapshotId);
-        } else {
-          // validate(TableMetadata) is called in apply(TableMetadata) after this apply refreshes the table state
-          return super.apply();
-        }
-
-      case ROLLBACK:
-        return base.snapshot(targetSnapshotId);
-
-      default:
-        throw new ValidationException("Invalid SnapshotManagerOperation: only cherrypick, rollback are supported");
-    }
-  }
-
-  private static void validateCurrentSnapshot(TableMetadata meta, Long requiredSnapshotId) {
-    if (requiredSnapshotId != null && meta.currentSnapshot() != null) {
-      ValidationException.check(meta.currentSnapshot().snapshotId() == requiredSnapshotId,
-          "Cannot fast-forward to non-append snapshot; current has changed: current=%s != required=%s",
-          meta.currentSnapshot().snapshotId(), requiredSnapshotId);
-    }
+    return transaction.table().currentSnapshot();
```

Comment:
Yeah, this is a bit weird. I thought about changing the API to implement `PendingUpdate<Map<String, SnapshotRef>>` so that this returns the refs map... but that's a breaking change and it doesn't seem worth it. With the current implementation, it will return at least the current snapshot of the main branch and that seems like reasonable validation.